### PR TITLE
handle and test a few unlikely cases when Spawners reuse tokens

### LINKED ISSUE #1343: hub doesn't check if the reused token is valid
This is a follow-up of a discussion with @minrk in gitter.

Observed when starting a container using `DockerSpawner` that failed on changing permissions due to mounting a read-only volume in `/home/jovyan` using hub 0.7.2 and dockerspawner 0.7.0.

The initial spawn failed to start, that triggered token removal since the spawn failed, but on the other hand the container was still reused by the spawner and still used the old token.

If I understand @minrk's explanation correctly, the hub should check that the reused hub token is still valid.

### PR DESCRIPTION
**Pull request by minrk**
- test that .will_resume preserves tokens (worked, but wasn't tested)

If a Spawner reuses a token, validate it in the db:

- verify that it's in the db
- if it doesn't map onto the right user, revoke the token
- if it's not in the db, insert it as a user-provided token

The most likely case is prior unclean shutdown of something like DockerSpawner, where a spawn failed and thus the token was revoked from the db, but the container was in fact created
and the token-reuse logic reused the defunct token.

closes #1343

### CODE REVIEW

File: jupyterhub/tests/test_spawner.py [hunk 1]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    assert user.spawner is spawner
+    # first start: gets a new API token
+    yield user.spawn()
+    api_token = spawner.api_token
+    found = orm.APIToken.find(app.db, api_token)
+    assert found
+    assert found.user.name == user.name
+    assert user.api_tokens == [found]
+    yield user.stop()
+    yield user.spawn()
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: I would suggest putting a comment before the second spawn
```# second spawn```

File: jupyterhub/tests/test_spawner.py [hunk 2]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    # verify re-use of API token
+    assert spawner.api_token == api_token
+    # verify that a new token was not created
+    assert user.api_tokens == [found]
+
+
+@pytest.mark.gen_test
+def test_spawner_insert_api_token(db, app):
+    user = add_user(app.db, app, name='tonkee')
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: ```# setup: add a user to db; check that no single user server running; no api_tokens exist for user```

File: jupyterhub/tests/test_spawner.py [hunk 3]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    if user.running:
+        yield user.stop()
+    assert user.api_tokens == []
+
+    spawner = user._new_spawner(name='', spawner_class=MockAPITokenSpawner)
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: `# setup: set up a spawner for the user`

File: jupyterhub/tests/test_spawner.py [hunk 4]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    user.spawners[''] = spawner
+    assert user.spawner is spawner
+
+    api_token = new_token()
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: `# test: generate a new token; test token is not in the db`

File: jupyterhub/tests/test_spawner.py [hunk 5]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    assert not orm.APIToken.find(app.db, api_token)
+    user.spawner.use_this_api_token = api_token
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: `# set user's spawner (the mock spawner in the test) to use the newly generated token which is not in the db

File: jupyterhub/tests/test_spawner.py [hunk 6]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    # Spawner's provided API token is inserted into the db
+    # This shouldn't happen unless there are bugs elsewhere (in the Spawner),
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: `# Under normal circumstances, the spawner's provided API token would not be inserted into the db unless there is a bug somewhere else (in the Spawner)`

File: jupyterhub/tests/test_spawner.py [hunk 7]
```diff
@@ -270,3 +273,90 @@ def stop():
...
+    yield user.stop()
+    yield user.spawn()
+    # verify re-use of API token
+    assert spawner.api_token == api_token
+    # verify that a new token was not created
+    assert user.api_tokens == [found]
+
+
+@pytest.mark.gen_test
+def test_spawner_insert_api_token(db, app):
```
Comment (willingc) [inline jupyterhub/tests/test_spawner.py]: @minrk this test was confusing to me

Perhaps add a docstring
`"""Token provided by spawner is not in the db. Insert token into db as a user-provided token."""`

File: jupyterhub/tests/test_spawner.py (hunk 6 above)
Comment (minrk) [inline jupyterhub/tests/test_spawner.py]: Thanks! should be better, now.

**Comment by minrk** [comment]
Thanks for the review. I think it should be clearer (and simpler) now.

**Pull request merged**